---
title: "scATACseq_merge_ClustersInvestigation"
author: "Laëtitia Racine"
date: "2022-09-06"
subtitle: "Dernière modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
---

```{r, Setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

```

```{r, Dependencies, warning=F, message=F}

library(Seurat)
library(Signac)

```

```{r, Working directories and external script}

directory = "/home/rparmentier/Bureau/Git_Metabo_Analysis/"
directory_exp = paste0(directory, "exp/")
directory_data = paste0(directory, "data/scATACseq/")
directory_bin = paste0(directory, "bin/")

# Create a unique folder for output corresponding to the date of the day
current_date = format(Sys.time(), "%Y%m%d")
dir.create(path = paste0(directory_exp,"scATACseq_merge_ClustersInvestigation/"))
dir.create(path = paste0(directory_exp,"scATACseq_merge_ClustersInvestigation/", current_date))
directory_output = paste0(directory_exp, "scATACseq_merge_ClustersInvestigation/", current_date, "/")

# Load external script with functions and constantes 
source(file = paste0(directory_bin, "functions_constantes.R"))

```

```{r, Input loading}

dir = pic_last_dir(paste0(directory_exp, "scATACseq_merge_NormalizationReduction/"))
seurat_merged = readRDS(paste0(dir, "/", "seurat_common_merged_qc_normalized_reduced.rds"))

```

https://github.com/satijalab/seurat/issues/738

```{r, Extract cell numbers per cluster for res 0.06, eval = FALSE}

df_clust_res0.06 = table(seurat_merged@meta.data$ATAC_combined_snn_res.0.06, seurat_merged@meta.data$orig.ident)
df_clust_res0.06

```

```{r, Extract cell numbers per cluster for res 0.25}

df_clust_res0.25 = table(seurat_merged@meta.data$ATAC_combined_snn_res.0.25, seurat_merged@meta.data$orig.ident)
df_clust_res0.25

```










# .Investigation sur les clusters obtenus 

```{r, Change DefaultAssay and choose resolution}

# Change back to working with peaks instead of gene activities
DefaultAssay(seurat_obj) <- 'peaks'

# Choose working resolution
res = "res.0.05"
meta_col = paste0("peaks_snn_", res)
Idents(seurat_obj) = (seurat_obj@meta.data)[,meta_col]

```

```{r, Subset seurat_obj based on cluster, eval = FALSE}

# Subset seurat obj to create one object per cluster
for (i in 1:length(levels((seurat_obj@meta.data)[,meta_col])) ) {

  # Extract cluster number
  clust = levels((seurat_obj@meta.data)[,meta_col])[i]
  print(paste("Cells extraction for clust ", clust))
  # Create subset of seurat object with cells from clust x
  expr = FetchData(object = seurat_obj, vars = meta_col)
  seurat_clust = seurat_obj[, which(expr == clust)]                       # ligne qui prend du temps
  # Check if the subset was correctly done
  print(table((seurat_clust@meta.data)[,meta_col]))

  # Extract peak count matrix of clust x from general seurat object
  print("Matrix count extraction")
  tab_clust = as.data.frame(GetAssayData(seurat_clust, slot = "data"))    # ligne qui prend du temps

  # Save object
  saveRDS(tab_clust, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_matrix_count_clust", clust, ".rds"))
  saveRDS(seurat_clust, file = paste0(dir_output, currentDate, "/seurat_clust", clust, "_dim", dims_choice_char, "_", res, ".rds"))

}

```


```{r, Peaks distribution per cell across clusters - tabs creation}

# https://github.com/satijalab/seurat/issues/1053

# Create dataframe with summary of peaks per cell and/or per annotation
tab_distripeaks_clust = data.frame()
tab_bulkpeaks_clust = data.frame()
  
for ( i in 1:length(levels((seurat_obj@meta.data)[,meta_col])) ) {
  
  # Extract cluster number
  clust = levels((seurat_obj@meta.data)[,meta_col])[i]
  print(paste("Cells extraction for clust ", clust))
  # Load subset of seurat object with cells from clust x and corresponding matrix count
  seurat_clust = readRDS(paste0(dir_output, currentDate, "/seurat_clust", clust, "_dim", dims_choice_char, "_", res, ".rds"))
  print("Load matrix count")
  tab_clust = readRDS(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_matrix_count_clust", clust, ".rds"))
  # Transform matrix count into logical
  tab_clust = tab_clust %>% mutate_all(as.logical)
  
  # Extract peaks_name according to genomic annotation
  print("Calcul peaks vs genomic annotation")
  peaks_name_list = list(
    total_peaks = seurat_clust@assays$peaks@ranges$peaks_name,
    peaks_FANTOM5_promoter = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$FANTOM5_promoter == TRUE])$peaks_name,
    peaks_UTR3P = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$UTR3P == TRUE])$peaks_name,
    peaks_UTR5P = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$UTR5P == TRUE])$peaks_name,
    peaks_CpG = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CpG == TRUE])$peaks_name,
    peaks_CTCF = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF == TRUE])$peaks_name,
    peaks_Exons = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$Exons == TRUE])$peaks_name,
    peaks_Introns = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$Introns == TRUE])$peaks_name,
    peaks_TSS_mp1kb = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$TSS_mp1kb == TRUE])$peaks_name,
    peaks_Intergenic = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$Intergenic == TRUE])$peaks_name,
    peaks_CpG_Intergenic = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CpG_Intergenic == TRUE])$peaks_name,
    peaks_CTCF_Intergenic = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF_Intergenic == TRUE])$peaks_name,
    peaks_CTCF_in_intron = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF_in_intron == TRUE])$peaks_name,
    peaks_CTCF_in_exon = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF_in_exon == TRUE])$peaks_name
    )

  # Calculate number of peaks total per annotation and per cell per annotation
  temp_distripeaks = lapply(peaks_name_list, function(x) {
    tab_filtered = tab_clust %>% dplyr::filter(row.names(tab_clust) %in% unlist(x))
    tab_calc = data.frame(colSums(tab_filtered)) 
    tab_calc = tab_calc %>% tibble::rownames_to_column(var = "cell")})
  temp_distripeaks = bind_rows(temp_distripeaks, .id = "peaks_category") %>% dplyr::mutate(clust = clust)
  colnames(temp_distripeaks) = c("peaks_category", "cell", "total_nbpeaks_percell", "clust")
  
  temp_bulkpeaks = data_frame(annotation = names(peaks_name_list), total_nbpeaks = as.numeric(lapply(peaks_name_list, length))) %>% 
    dplyr::mutate(clust = clust, .before = "total_nbpeaks")
  
  # Bind in general tab
  tab_bulkpeaks_clust = rbind(tab_bulkpeaks_clust, temp_bulkpeaks)
  tab_distripeaks_clust = rbind(tab_distripeaks_clust, temp_distripeaks)

}

# Add statistical information
tab_distripeaks_stat = tab_distripeaks_clust %>%
  dplyr::group_by(clust, peaks_category) %>%
  dplyr::mutate(nbcells_clust = n(),
                median_nbpeaks_percell = median(total_nbpeaks_percell),
                mean_nbpeaks_percell = round(mean(total_nbpeaks_percell)),
                max_nbpeaks_percell = max(total_nbpeaks_percell),
                min_nbpeaks_percell = min(total_nbpeaks_percell),
                var_nbpeaks_percell = round(sd(total_nbpeaks_percell))) %>%
  dplyr::mutate(percent_cell_clust = round((nbcells_clust/ncol(seurat_obj))*100)) %>%
  dplyr::ungroup() %>%
  dplyr::select(peaks_category, clust, nbcells_clust, percent_cell_clust, median_nbpeaks_percell, 
                mean_nbpeaks_percell, max_nbpeaks_percell, min_nbpeaks_percell, var_nbpeaks_percell) %>%
  dplyr::distinct()

# Save dataframes and clean working space
write_csv2(tab_bulkpeaks_clust, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_bulkpeaks.csv"))
write_csv2(tab_distripeaks_clust, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_distripeaks.csv"))
write_csv2(tab_distripeaks_stat, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_distripeaks_stat.csv"))
rm(temp_bulkpeaks, temp_distripeaks, peaks_name_list, clust, seurat_clust, i, expr, tab_clust)

```

```{r, Peaks distribution per cell across clusters - visualization, results = 'hold', fig.width=20, fig.height=6}

# load files (facultatif)
# tab_bulkpeaks_clust = read_csv2(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_bulkpeaks.csv"))
# tab_distripeaks_clust = read_csv2(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_distripeaks.csv"))
# tab_distripeaks_stat = read_csv2(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_distripeaks_stat.csv"))

# Visualize output tab
tab_distripeaks_clust %>% 
  kable(caption = res) %>% 
  kable_styling()  %>%
  scroll_box(height = "250px", width = "100%")

tab_distripeaks_stat %>% 
  kable(caption = res) %>% 
  kable_styling()  %>%
  scroll_box(height = "250px", width = "100%")

tab_bulkpeaks_clust %>%
  kable(caption = res) %>% 
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")

# Histogram : nbpeaks per annotation per cluster
plot_bulkpeak = ggplot(tab_bulkpeaks_clust, aes(x=annotation, y=total_nbpeaks, fill=clust)) +
  geom_bar(stat="identity", position=position_dodge()) +
  ylab(label = "TOTAL PEAK NUMBER IN CLUSTER") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank())
plot_bulkpeak

# Distribution : nbpeaks per cell per annotation 
tab_plot  = full_join(tab_distripeaks_clust, tab_distripeaks_stat, by = c("peaks_category", "clust"))
tab_plot$clust = as.factor(tab_plot$clust)
plot_distripeak = ggplot(tab_plot, aes(x=total_nbpeaks_percell, fill=clust)) +
  geom_histogram(bins = 50, color = "black", alpha = 0.5) +
  geom_vline(aes(xintercept = median_nbpeaks_percell, col=clust), linetype = "dashed", size = 1) +
  ggtitle(label = paste0("Peaks' distribution per cell according to cluster (with median) - ", res, " !!! Different scales !!!")) +
  ylab(label = "CELL COUNT") +
  xlab("NBPEAKS PER CELL") +
  facet_wrap(facets = .~peaks_category, scales = "free") 
plot_distripeak

# Save objects
ggsave(plot = plot_bulkpeak,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_hist_peak_annot_clust.png"),
       width = 14, height = 6)
ggsave(plot = plot_distripeak,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_distri_peak_annot_clust.png"),
       width = 16, height = 10)

# Save plot for article 
plot_article_total = ggplot(tab_plot %>% filter(peaks_category == "total_peaks"), aes(x=total_nbpeaks_percell, fill=clust)) +
  geom_histogram(bins = 50, color = "black", alpha = 0.5) +
  geom_vline(aes(xintercept = median_nbpeaks_percell, col=clust), linetype = "dashed", size = 1) +
  ylab(label = "CELL COUNT") +
  xlab(label = "NBPEAKS PER CELL") +
  ggtitle(label = "TOTAL PEAKS")
plot_article_prom = ggplot(tab_plot %>% filter(peaks_category == "peaks_FANTOM5_promoter"), aes(x=total_nbpeaks_percell, fill=clust)) +
  geom_histogram(bins = 50, color = "black", alpha = 0.5) +
  geom_vline(aes(xintercept = median_nbpeaks_percell, col=clust), linetype = "dashed", size = 1) +
  ylab(label = "CELL COUNT") +
  xlab(label = "NBPEAKS PER CELL") +
  ggtitle(label = "PEAKS SPANNING PROMOTERS")
plot_article_intergenic = ggplot(tab_plot %>% filter(peaks_category == "peaks_Intergenic"), aes(x=total_nbpeaks_percell, fill=clust)) +
  geom_histogram(bins = 50, color = "black", alpha = 0.5) +
  geom_vline(aes(xintercept = median_nbpeaks_percell, col=clust), linetype = "dashed", size = 1) +
  ylab(label = "CELL COUNT") +
  xlab(label = "NBPEAKS PER CELL") +
  ggtitle(label = "PEAKS FROM INTERGENIC REGION")

ggsave(plot = plot_article_total | plot_article_prom | plot_article_intergenic,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_distri_peak_annot_clust_article.pdf"),
       width = 16, height = 4)

```

```{r, Peaks distribution per cell across clusters - Focus Clust0_1 Inter_Prom, results = 'hold', fig.width=20, fig.height=10}

# Calculate difference between median nbpeaks per cell in clust 1 and in clust 0 for each annotation category
tab_delta = tab_plot %>% 
  dplyr::select(peaks_category, clust, median_nbpeaks_percell) %>% 
  dplyr::filter(clust %in% c("0", "1")) %>%
  dplyr::distinct() %>%
  dplyr::group_by(peaks_category) %>%
  dplyr::mutate(diff_clust1_clust0 = median_nbpeaks_percell[clust == "0"] - median_nbpeaks_percell[clust == "1"])

tab_delta %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(height = "250px", width = "100%")
  
# Visualize plot woth only promoter and intergenic
tab_inter_prom = tab_plot %>% dplyr::filter(peaks_category %in% c("peaks_FANTOM5_promoter", "peaks_Intergenic"))
plot_distri_interprom = ggplot(tab_inter_prom, aes(x=total_nbpeaks_percell, fill=clust)) +
  geom_histogram(bins = 50, color = "black", alpha = 0.5) +
  geom_vline(aes(xintercept = median_nbpeaks_percell, col=clust), linetype = "dashed", size = 1) +
  ggtitle(label = paste0("Peaks' distribution per cell according to cluster (with median) - ", res)) +
  ylab(label = "CELL COUNT") +
  xlab("NBPEAKS PER CELL") +
  facet_wrap(facets = .~peaks_category)
plot_distri_interprom

# Save object and clean working space
ggsave(plot = plot_distri_interprom,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_distri_peak_annot_clust_zoominterprom.png"),
       width = 16, height = 10)
write_csv2(tab_delta, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_delta_clust1_clust0.csv"))
rm(tab_bulkpeaks_clust, tab_distripeaks_clust, tab_distripeaks_stat, tab_plot, plot_distripeak, plot_bulkpeak, tab_inter_prom, tab_delta, plot_distri_interprom)

```

```{r, Verification reads number}

# somme des reads : est-ce que toutes les cellules versus cellules clust0 versus cellules clust1 ont bien un nombre de reads différents pour les peaks ?

# Extract matrix count for each cells 
tab_all = as.data.frame(GetAssayData(seurat_obj, slot = "data")) 
tab_clust0 = readRDS(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_matrix_count_clust0.rds"))
tab_clust1 = readRDS(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_matrix_count_clust1.rds"))
# On a bien trois tableaux avec le même nombre de lignes (peaks) mais qui diffèrent par leur nombre de colonnes (cellules)

# Compare matrix count
tab_compare = data.frame(nbreads_allcells = rowSums(tab_all),
                         nbreads_clust0 = rowSums(tab_clust0),
                         nbreads_clust1 = rowSums(tab_clust1))  %>% 
  tibble::rownames_to_column(var = "peak_name")

# Save object and clean workspace 
saveRDS(tab_all, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_matrix_count.rds"))
write_csv2(tab_compare, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_nbreads_clust.csv"))
rm(tab_all, tab_clust0, tab_clust1, tab_compare)

```

```{r, Sharing peaks in cell across clusters, results = 'hold', fig.width=6, fig.height=16}

tab_districell_clust = data.frame()

for ( i in 1:length(levels((seurat_obj@meta.data)[,meta_col])) ) {
  
  # Extract cluster number
  clust = levels((seurat_obj@meta.data)[,meta_col])[i]
  print(paste("Cells extraction for clust ", clust))
 # Load seurat object subset on clust x 
  seurat_clust = readRDS(paste0(dir_output, currentDate, "/seurat_clust", clust, "_dim", dims_choice_char, "_", res, ".rds"))
 # Load matrix count for cells from clust x 
  tab_clust = readRDS(paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_matrix_count_clust", clust, ".rds"))
  tab_clust = tab_clust %>% mutate_all(as.logical)
  
  # Extract peaks_name according to genomic annotation
  peaks_name_list = list(
    total_peaks = seurat_clust@assays$peaks@ranges$peaks_name,
    peaks_FANTOM5_promoter = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$FANTOM5_promoter == TRUE])$peaks_name,
    peaks_UTR3P = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$UTR3P == TRUE])$peaks_name,
    peaks_UTR5P = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$UTR5P == TRUE])$peaks_name,
    peaks_CpG = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CpG == TRUE])$peaks_name,
    peaks_CTCF = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF == TRUE])$peaks_name,
    peaks_Exons = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$Exons == TRUE])$peaks_name,
    peaks_Introns = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$Introns == TRUE])$peaks_name,
    peaks_TSS_mp1kb = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$TSS_mp1kb == TRUE])$peaks_name,
    peaks_Intergenic = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$Intergenic == TRUE])$peaks_name,
    peaks_CpG_Intergenic = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CpG_Intergenic == TRUE])$peaks_name,
    peaks_CTCF_Intergenic = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF_Intergenic == TRUE])$peaks_name,
    peaks_CTCF_in_intron = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF_in_intron == TRUE])$peaks_name,
    peaks_CTCF_in_exon = (seurat_clust@assays$peaks@ranges[seurat_clust@assays$peaks@ranges$CTCF_in_exon == TRUE])$peaks_name
    )

  # Calculate number of peaks total per annotation and per cell per annotation
 temp_districell_clust = lapply(peaks_name_list, function(x) {
    tab_filtered = tab_clust %>% dplyr::filter(row.names(tab_clust) %in% unlist(x))
    tab_calc = data.frame(rowSums(tab_filtered)) 
    tab_calc = tab_calc %>% tibble::rownames_to_column(var = "peak")})
  temp_districell_clust = bind_rows(temp_districell_clust, .id = "peaks_category") %>% dplyr::mutate(clust = clust, nbcells_clust = ncol(tab_clust))
  colnames(temp_districell_clust) = c("peaks_category", "peak", "nbcells_sharing_peak", "clust", "nbcells_clust")
  temp_districell_clust = temp_districell_clust %>% dplyr::mutate(propcells_sharing_peak = round((nbcells_sharing_peak/nbcells_clust)*100))
  tab_districell_clust = rbind(tab_districell_clust, temp_districell_clust)
  
}

tab_districell_stat = tab_districell_clust %>%
  dplyr::group_by(clust, peaks_category) %>%
  dplyr::mutate(median_nbcells_sharing_peak = median(nbcells_sharing_peak),
                mean_nbcells_sharing_peak = round(mean(nbcells_sharing_peak)),
                max_nbcells_sharing_peak = max(nbcells_sharing_peak),
                min_nbcells_sharing_peak = min(nbcells_sharing_peak),
                var_nbcells_sharing_peak = round(sd(nbcells_sharing_peak)),
                median_propcells_sharing_peak = median(propcells_sharing_peak),
                mean_propcells_sharing_peak = round(mean(propcells_sharing_peak)),
                max_propcells_sharing_peak = max(propcells_sharing_peak),
                min_propcells_sharing_peak = min(propcells_sharing_peak),
                var_propcells_sharing_peak = round(sd(propcells_sharing_peak))) %>%
  dplyr::ungroup() %>%
  dplyr::select(peaks_category, clust, median_nbcells_sharing_peak, mean_nbcells_sharing_peak,
                max_nbcells_sharing_peak, min_nbcells_sharing_peak, var_nbcells_sharing_peak,
                median_propcells_sharing_peak, mean_propcells_sharing_peak, max_propcells_sharing_peak, 
                min_propcells_sharing_peak, var_propcells_sharing_peak) %>%
  dplyr::distinct()

# Distribution : nbpeaks per cell per annotation 
tab_plot  = full_join(tab_districell_clust, tab_districell_stat, by = c("peaks_category", "clust"))
tab_plot$clust = as.factor(tab_plot$clust)

legend.grid = c("0" = "cluster 0", "1" = "cluster 1", "2" = "cluster 2", "3" = "cluster 3", "4" = "cluster 4")
plot_districell_grid = ggplot(tab_plot, aes(x=propcells_sharing_peak, fill=clust)) +
  geom_histogram(binwidth = 10, color = "black", alpha = 0.5, position = "identity") +
  geom_vline(aes(xintercept = median_propcells_sharing_peak, col=clust), linetype = "dashed", size = 1) +
  ggtitle(label = paste0("Dashed line correspond to median - ", res, " !!! Different scales !!!")) +
  ylab(label = "NUMBER OF PEAKS CONCERNED") +
  xlab("PROPORTION OF CELLS SHARING A PEAK IN THE CLUSTER") +
  facet_grid(peaks_category ~ clust, labeller = labeller(clust = legend.grid), scales = "free") +
  theme(legend.position = "none")
plot_districell_grid

plot_districell = ggplot(tab_plot, aes(x=propcells_sharing_peak, fill=clust)) +
  geom_histogram(binwidth = 10, aes(color = clust), alpha = 0.2, position = "identity") +
  geom_vline(aes(xintercept = median_propcells_sharing_peak, col=clust), linetype = "dashed", size = 1) +
  ggtitle(label = paste0("Dashed line correspond to median - ", res, " !!! Different scales !!!")) +
  ylab(label = "NUMBER OF PEAKS CONCERNED") +
  xlab("PROPORTION OF CELLS SHARING A PEAK IN THE CLUSTER") +
  facet_grid(peaks_category~., scales = "free") +
  theme(legend.position = "none")
plot_districell

plot_districell_density = ggplot(tab_plot, aes(x=propcells_sharing_peak, fill=clust)) +
  geom_density(aes(col=clust), alpha = 0.3) +
  geom_vline(aes(xintercept = median_propcells_sharing_peak, col=clust), linetype = "dashed", size = 1) +
  ggtitle(label = paste0("Dashed line correspond to median - ", res, " !!! Different scales !!!")) +
  ylab(label = "PROPOTION OF PEAKS CONCERNED") +
  xlab("PROPORTION OF CELLS SHARING A PEAK IN THE CLUSTER") +
  facet_grid(peaks_category~., scales = "free") +
  theme(legend.position = "none")
plot_districell_density

# Save object and clean working space 
ggsave(plot = plot_districell,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_distri_cell_annot_clust.png"),
       width = 6, height = 14)
ggsave(plot = plot_districell_grid,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_distrigrid_cell_annot_clust.png"),
       width = 16, height = 16)
ggsave(plot = plot_districell_density,
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_distridensity_cell_annot_clust.png"),
       width = 6, height = 14)
write_csv2(tab_districell_clust, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_districell.csv"))
write_csv2(tab_districell_stat, file = paste0(dir_output, currentDate, "/tab_dim", dims_choice_char, "_", res, "_districell_stat.csv"))
rm(tab_plot, tab_districell_stat, tab_districell_clust, temp_districell_clust, peaks_name_list, clust, tab_clust, seurat_clust, i)
rm(plot_districell_density, plot_districell_grid, plot_districell, legend.grid)

```

```{r, Isolate cluster specific peak}

# sortir la liste de tous les gènes qui ont un peak dans le clust0, dans le clust1, enlever l'intersection et regarder ce qui n'est pas en commun :
  # - sous sampling que les peaks qui tombent dans TSS (refaire FindMarkers ?)
  # - utiliser closest feature sur tous ces peaks par cluster
  # - faire l'exclusion pour sortir la liste particulière de chaque cluster
  # - GO sur les deux listes

# Subset of seurat object to keep only cells from clust0
clust = "0"
seurat_clust0 = readRDS(paste0(dir_output, currentDate, "/seurat_clust", clust, "_dim", dims_choice_char, "_", res, ".rds"))
clust0_gene <- ClosestFeature(seurat_clust0, regions = granges(seurat_clust0)) 
clust0_gene_list = as.vector(clust0_gene$gene_name)

# Subset of seurat object to keep only cells from clust1
clust = "1"
seurat_clust1 = readRDS(paste0(dir_output, currentDate, "/seurat_clust", clust, "_dim", dims_choice_char, "_", res, ".rds"))
clust1_gene <- ClosestFeature(seurat_clust1, regions = granges(seurat_clust1))
clust1_gene_list = as.vector(clust1_gene$gene_name)

# pb : tous les gènes sont représentés dans les deux clusters, il faut s'intéresser aux proportions d'accessibilité
table(clust0_gene_list == clust1_gene_list)

rm(seurat_clust0, clust0_gene, clust0_gene_list, seurat_clust1, clust1_gene, clust1_gene_list, clust)
```

```{r, Isolate cluster specific top 200 peaks - Gene analysis}

# On isole les peaks du clust0 et du clust1 qui sont partagés par le plus de cellules dans le cluster et on compare les listes obtenues 

########### TRAVAIL A PARTIR DES PEAKS DES PROMOTEURS #######################
# Attention !!! On applique ici le filtre sur les peaks pour garder que les promoteurs avant de sélectionner les 200 peaks les plus partagés
# Si on le fait après, ça change beaucoup le nombre de peaks à extraire

# Load (facultatif)
tab_districell_clust = read.csv2(paste0(dir_output, currentDate, "/tab_dim2-30_res.0.05_districell.csv"))

tab_clust0 = tab_districell_clust %>% 
  dplyr::filter(clust == "0") %>%
  dplyr::filter(peaks_category == "peaks_FANTOM5_promoter") %>% 
  dplyr::arrange(desc(nbcells_sharing_peak))  %>%
  dplyr::select(-clust)
# Tab contenant les 200 peaks les plus partagés par les cellules du cluster0
clust0_top200 = head(tab_clust0, 200)
clust0_top200 = as.vector(clust0_top200$peak)

tab_clust1 = tab_districell_clust %>% 
  dplyr::filter(clust == "1") %>%
  dplyr::filter(peaks_category == "peaks_FANTOM5_promoter") %>% 
  dplyr::arrange(desc(nbcells_sharing_peak)) %>%
  dplyr::select(-clust)
# Tab contenant les 200 peaks les plus partagés par les cellules du cluster1
clust1_top200 = head(tab_clust1, 200)
clust1_top200 = as.vector(clust1_top200$peak)

# Comparer les deux listes de peaks obtenues 
table(clust0_top200 == clust1_top200) # 4 peaks en commun et 196 peaks différents 

# Extraire les peaks qui ne sont présents que dans le clust0 ou que dans le clust1
clust0_top200_spe = clust0_top200[(clust0_top200 == clust1_top200) != TRUE]
clust1_top200_spe = clust1_top200[(clust0_top200 == clust1_top200) != TRUE]

# Utiliser closest feature sur tous ces peaks par cluster
gene_clust0 = ClosestFeature(seurat_obj, regions = clust0_top200_spe)
gene_clust1 = ClosestFeature(seurat_obj, regions = clust1_top200_spe)

# GO clust0
name_correspondance = bitr(gene_clust0$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")
colnames(gene_clust0)[3] = "ENSEMBL"
gene_clust0 = inner_join(gene_clust0, name_correspondance, by = "ENSEMBL")
ego_clust0 <- enrichGO(gene = gene_clust0$ENTREZID,
       OrgDb         = org.Hs.eg.db,
       ont           = "ALL", # ont = ALL => cellular components, cellular pathway...
       pAdjustMethod = "BH",
       pvalueCutoff  = 0.05,
       qvalueCutoff  = 0.2,
       readable      = TRUE)
go_200peaks_clust0 = dotplot(ego_clust0, showCategory = 20)
go_200peaks_clust0

# GO clust1
name_correspondance = bitr(gene_clust1$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")
colnames(gene_clust1)[3] = "ENSEMBL"
gene_clust1 = inner_join(gene_clust1, name_correspondance, by = "ENSEMBL")
ego_clust1 <- enrichGO(gene = gene_clust1$ENTREZID,
       OrgDb         = org.Hs.eg.db,
       ont           = "ALL", # ont = ALL => cellular components, cellular pathway...
       pAdjustMethod = "BH",
       pvalueCutoff  = 0.05,
       qvalueCutoff  = 0.2,
       readable      = TRUE)
go_200peaks_clust1 = dotplot(ego_clust1, showCategory = 20)
go_200peaks_clust1

# Save and clean working space 
ggsave(plot = go_200peaks_clust0, 
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_go_top200peaks_clust0.png"),
       width = 6, height = 6)
ggsave(plot = go_200peaks_clust1, 
       filename = paste0(dir_output, currentDate, "/plot_dim", dims_choice_char, "_", res, "_go_top200peaks_clust1.png"),
       width = 6, height = 6)
rm(clust0_top200, clust0_top200_spe, clust1_top200, clust1_top200_spe, tab_clust0, tab_clust1)
rm(ego_clust0, ego_clust1, gene_clust0, gene_clust1, name_correspondance, go_200peaks_clust0, go_200peaks_clust1)
```