---
title: "scATACseq_ind_CreateSeuratObject"
author: "Laëtitia Racine"
date: "2022-09-22"
subtitle: "Dernière modification : `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: "hide"
    toc: true
    toc_float: false
    theme: journal
---

<style>
body {text-align: justify}
</style>

```{r, Setup, include=F}

knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      # record the current time before each chunk
      now <<- Sys.time()
    } else {
      # calculate the time difference after a chunk
      res <- difftime(Sys.time(), now)
      # return a character string to show the time
      paste("Time for this code chunk to run:", res)
    }
  }
}))

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(time_it = TRUE)


```

```{r, Link with Snakefile}

"Create Seurat Object from Cell Ranger outputs

Usage:
  scATACseq_Snakemake_ind_CreateSeurat_Object.Rmd [options] <matrixh5_file> <fragment_file> <metadata_file> <output>
  scATACseq_Snakemake_ind_CreateSeurat_Object.Rmd -h | --help

Options:
  -h, --help               Show this screen
" -> doc

library(docopt)
arguments <- docopt(doc)

```

```{r, Dependencies}

library(Signac)
library(Seurat)
library(stringr)
library(EnsDb.Hsapiens.v86)   # v.86 = hg38 | v.75 = hg19
library(GenomeInfoDb)
library(knitr)

```

```{r, Input loading}

matrixh5 = arguments$matrixh5_file
fragment = arguments$fragment_file
singlecell = arguments$metadata_file

name_cond = str_extract(string = singlecell, pattern = "[:alnum:]+(?=_singlecell.csv)")

```

```{r, Create seurat object}

counts <- Seurat::Read10X_h5(filename = matrixh5)

metadata <- read.csv(
    file = singlecell,
    header = TRUE,
    row.names = 1
  )

chrom_assay <- CreateChromatinAssay(
    counts = counts,
    sep = c(":", "-"),
    genome = 'hg38',
    fragments = fragment,
    min.cells = 100,
    min.features = 200
  )

seurat_obj <- CreateSeuratObject(
    counts = chrom_assay,
    assay = "peaks",
    meta.data = metadata)


```


```{r, Add annotations}

# Extract annotations from EnsDb database
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86, standard.chromosomes = TRUE)
seqlevelsStyle(annotations) = "UCSC"
print(table(annotations$type))

# Add annotation database into seurat object (needed for downstream analysis (QC functions))
Annotation(seurat_obj) = annotations

# Visualize first object of the list to see the annotations
seurat_obj@assays$peaks@annotation

```

```{r, Update orig.ident}

Idents(object = seurat_obj) = name_cond
seurat_obj$orig.ident = name_cond

```

```{r, Save output}

saveRDS(object = seurat_obj, file = arguments$output)

```

```{r, Rsession}

# Clean working space and show package version
rm(list = ls())
sessionInfo()

```
